### PR TITLE
feat: extract deploy logic from js-client

### PR DESCRIPTION
**Pull request by eduardoboucas**
**- Summary**

Extracts deploy logic from `netlify/js-client` into the CLI core. Part of https://github.com/netlify/js-client/issues/157.

**- Test plan**

Copied across the unit tests from `netlify/js-client`.

**- A picture of a cute animal (not mandatory but encouraged)**

8d303e65cc308b08efd2bd3430a6ce5c

### CODE REVIEW

File: src/utils/deploy/hash-files.test.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
+const path = require('path')
+
+const test = require('ava')
+
+const { DEFAULT_CONCURRENT_HASH } = require('./constants')
+const { hashFiles } = require('./hash-files')
+
+test('Hashes files in a folder', async (t) => {
+  const dirname = path.resolve(__dirname, '../../../tests/site-with-functions')
```
Comment (erezrokah) [inline src/utils/deploy/hash-files.test.js:9]: [sand] Can we use the `site-builder` https://github.com/netlify/cli/blob/e9717ff8aa3c52c2a3d7a95f4f86bbb1f5b79668/tests/utils/site-builder.js to keep everything related to the test in the same file?

**Review by erezrokah — APPROVED** [review]
🚀

File: src/utils/deploy/hash-files.test.js (hunk 1 above)
Comment (eduardoboucas) [reply src/utils/deploy/hash-files.test.js:9]: Absolutely, great point. I'll update.

Comment (eduardoboucas) [reply src/utils/deploy/hash-files.test.js]: Done in https://github.com/netlify/cli/pull/2317/commits/af3925a3a449be20fb1766ecd7a31bb273365441.

File: src/utils/deploy/hash-files.test.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      })
+      .withContentFile({
+        path: 'functions/function-1.js',
+        content: `
+const { one } = require('../../lib/util')
+
+module.exports.handler = async () => ({ statusCode: 200, body: one })
+`,
+      })
+      .withContentFile({
```
Comment (erezrokah) [inline src/utils/deploy/hash-files.test.js:27]: [dust] The site builder has `withFunction`, also it has `withContentFiles` that supports an array

**Review by erezrokah — APPROVED** [review]

File: src/utils/deploy/hash-files.test.js (hunk 2 above)
Comment (eduardoboucas) [reply src/utils/deploy/hash-files.test.js]: I know about `withFunction`, but I wanted to `require` the util file which I wouldn't normally do inside the handler, which is why I've opted for using `withContentFile`.

In https://github.com/netlify/cli/pull/2317/commits/008a66b491b4b5c2aeecfd5c6ac7ae188cebe1e6 I've changed it to `withContentFiles`.

**Review by erezrokah — APPROVED** [review]

**Comment by eduardoboucas** [comment]
@erezrokah I have pushed one last to this PR in https://github.com/netlify/cli/pull/2317/commits/735bb19c4030813a257a7f467410073c656b573c, where I'm passing the functions configuration to the `zip-it-and-ship-it` call. This was the main reason for me to move this logic to the CLI, since currently function options (including the bundler choice) are not being respected in the `deploy` command.

If this looks good to you, feel free to release this on Sunday. Thanks! 🙌🏻

**Review by erezrokah — APPROVED** [review]

**Pull request merged**